File: ProjectEuler/301-350/S347.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ProjectEuler
{
    public class S347 : ISolution
    {
        private static readonly int MAX = 10000000;

        public string GetAnswer()
        {
            var Ms = new List<long>();

            var primes = Utils.GenPrimes(MAX / 2);
            for (int i = 0; ; i++)
            {
                if (primes[i] == 0)
                {
                    continue;
                }

                if (primes[i] > Math.Sqrt(MAX))
                {
                    break;
                }

                for (int j = i + 1; j < primes.Length; j++)
                {
                    if (primes[j] == 0)
                    {
                        continue;
                    }

                    long p = primes[i];
                    long q = primes[j];
                    if (p * q < MAX)
                    {
                        Ms.Add(GetM(p, q));
                    }
                    else
                    {
                        break;
                    }
                }
            }

            return Ms.Sum().ToString(); ;
        }

        private static long GetM(long p, long q)
        {
            var max = 0L;
            var e1Max = (int)(Math.Log(MAX) / Math.Log(p));
            var e2Max = (int)(Math.Log(MAX) / Math.Log(q));
            for (int e1 = 0; e1 < e1Max; e1++)
            {
                for (int e2 = 0; e2 < e2Max; e2++)
                {
                    var product = Math.Pow(p, e1 + 1) * Math.Pow(q, e2 + 1);
                    if (product > MAX)
                    {
                        break;
                    }

                    if (product <= MAX && product > max)
                    {
                        max = (long)product;
                    }
                }
            }

            return max;
        }
    }
}

File: src/301-350/S347.cc
#include "S347.h"
#include "../Utils.h"
#include <algorithm>
#include <cmath>
#include <string>
#include <vector>

using namespace std;

namespace ProjectEuler
{
int64_t S347::GetM(int64_t p, int64_t q, int64_t MAX)
{
	int64_t best = 0;

	int64_t p_pow = p;
	while (p_pow * q <= MAX)
	{
		int64_t temp = p_pow * q;
		while (temp * q <= MAX)
		{
			temp *= q;
		}
		if (temp > best)
		{
			best = temp;
		}

		if (MAX / p < p_pow)
		{
			break;
		}
		p_pow *= p;
	}
	return best;
}

std::string S347::GetAnswer()
{
	int64_t MAX = 10000000;
	auto primes = Utils::GenPrimes(MAX / 2);

	int64_t sum = 0;
	// Optimization: if p*q > MAX, break inner.
	// If p*2 > MAX, break outer.

	// However, problem says sum of distinct M(p,q,N).
	// Since M(p,q,N) is divisible by p and q only, and <= N.
	// It is unique for the pair {p,q}.
	// And distinct pairs produce distinct numbers (fundamental theorem of arithmetic).
	// So we can sum them directly.

	for (size_t i = 0; i < primes.size(); ++i)
	{
		int64_t p = primes[i];
		if (p * p > MAX)
		{
			break; // Optimization: q > p, so p*q > p*p. If p*p > MAX, then p*q > MAX.
		}

		for (size_t j = i + 1; j < primes.size(); ++j)
		{
			int64_t q = primes[j];
			if (p * q > MAX)
			{
				break;
			}

			sum += GetM(p, q, MAX);
		}
	}

	return std::to_string(sum);
}
} // namespace ProjectEuler
